### PR TITLE
avoid calling get bucket object lock config for same bucket

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

File: cmd/client-s3.go [hunk 1]
```diff
@@ -1005,7 +1006,13 @@ func (c *S3Client) Put(ctx context.Context, reader io.Reader, size int64, metada
...
 	}
 	if opts.LegalHold == "" && opts.Mode == "" {
-		if _, _, _, err := c.api.GetBucketObjectLockConfig(ctx, bucket); err == nil {
+		_, ok := c.bucketObjectLockConfigCache.Load(bucket)
+		if !ok {
+			if _, _, _, err := c.api.GetBucketObjectLockConfig(ctx, bucket); err == nil {
+				c.bucketObjectLockConfigCache.Store(bucket, err == nil)
+				opts.SendContentMd5 = true
+			}
+		} else {
```
[the remark this answers is not in the document]
Comment (harshavardhana) [reply cmd/client-s3.go:1015]: @poornas this is only meant for success usecase i.e when object lock is set

**Comment by harshavardhana** [comment]
Re-implemented with a better mechanism PTAL @poornas @kannappanr @krisis

**Comment by poornas** [comment]
> Re-implemented with a better mechanism PTAL @poornas @kannappanr @krisis

When object locking is not enabled on the bucket, current implementation ends up doing a GetBucketObjectLockConfig call per cp operation. Even without that, trace on a recursive cp shows multiple invocations, probably because it is looking up per prefix

**Comment by harshavardhana** [comment]
> > Re-implemented with a better mechanism PTAL @poornas @kannappanr @krisis
> 
> When object locking is not enabled on the bucket, current implementation ends up doing a GetBucketObjectLockConfig call per cp operation. Even without that, trace on a recursive cp shows multiple invocations, probably because it is looking up per prefix

Looks like you didn't take the latest PR the call to `GetBucketObjectLockConfig` is only once in the beginning of the doCopySession() 

```
mc cp -r ~/data s3/tbucket13
```

```
mc: <DEBUG> GET /?object-lock= HTTP/1.1
Host: tbucket13.s3.dualstack.us-east-1.amazonaws.com                                           
User-Agent: MinIO (linux; amd64) minio-go/v7.0.5 mc/2020-08-25T01:29:14Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20200825/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855                                                                                                        
X-Amz-Date: 20200825T012942Z                                                                                                                                                                  
Accept-Encoding: gzip                                                                          
                                                                                                                                                                                              
mc: <DEBUG> HTTP/1.1 200 OK                                                                    
Transfer-Encoding: chunked                                                                                                                                                                    
Date: Tue, 25 Aug 2020 01:29:43 GMT                                                                                                                                                           Server: AmazonS3                                                                                                                                                                              
X-Amz-Id-2: VFCapH1fEUKD4qT5f0Lph1YOPKHbFaicU2ZokIcfHDIiV9+OputPuxoHBq7IhCmtEMDnWrMppas=
X-Amz-Request-Id: 719BAC0E4EB16EE3                                                                                                                                                            
                                                                                               
mc: <DEBUG> Response Time:  92.183632ms         ```

and subsequently, no other Put caller calls GetObjectLock()

Now calling this on a bucket which doesn't have object-locking

```
mc: <DEBUG> GET /?object-lock= HTTP/1.1                                                        
Host: harshavardhana.s3.dualstack.us-west-1.amazonaws.com                                                                                                                                     
User-Agent: MinIO (linux; amd64) minio-go/v7.0.5 mc/2020-08-25T01:29:14Z                       
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20200825/us-west-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**                        
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855         
X-Amz-Date: 20200825T013016Z                                                                   
Accept-Encoding: gzip                                                                                                                                                                         
                                                                                                                                                                                              
mc: <DEBUG> HTTP/1.1 404 Not Found                                                                                                                                                            
Transfer-Encoding: chunked                                                                     
Content-Type: application/xml                                                                  
Date: Tue, 25 Aug 2020 01:30:15 GMT                                                                                                                                                           
Server: AmazonS3                                                                                                                                                                              
X-Amz-Id-2: WzBliq96YNs0kqjfE+JP4TF+MjZ0YTNDuAKXN+vxNqawXXDRniHvwqKXlHFY388SYRCfwta01b8=                                                                                                      
X-Amz-Request-Id: 9F540F14BD9BD959                                                             
                                                                                               
15d                                                                                                                                                                                           
<?xml version="1.0" encoding="UTF-8"?>                                                         
<Error><Code>ObjectLockConfigurationNotFoundError</Code><Message>Object Lock configuration does not exist for this bucket</Message><BucketName>harshavardhana</BucketName><RequestId>9F540F14B
D9BD959</RequestId><HostId>WzBliq96YNs0kqjfE+JP4TF+MjZ0YTNDuAKXN+vxNqawXXDRniHvwqKXlHFY388SYRCfwta01b8=</HostId></Error>  ```

Sees 404 and no other caller calls

**Comment by harshavardhana** [comment]
So I don't see how it can happen that it calls for all objects unless you have a wrong `mc` binary locally.

**Comment by harshavardhana** [comment]
Even if the call is at a prefix, it's only once per destination

```
~ mc cp --debug -r /usr/sbin s3/tbucket13/prefix/
```

**Comment by poornas** [comment]
> So I don't see how it can happen that it calls for all objects unless you have a wrong `mc` binary locally.

yes, seem to have had a older binary. It is working ok.

**Pull request merged**